File: webgl/source/js/renders/lighttexturedcuberender.js
class LightTexturedCubeRender extends Render {


    matrix
    projMatrix
    viewModelMatrix
    normalMatrix

    prog
    idVao
    ids
    idTexture
    rotY = 0
    rotX = 0

    onSetup(appOGL) {
        super.onSetup(appOGL);
        const gl = appOGL.gl;
        const mat4 = glMatrix.mat4;
        const TO_RAD = Math.PI / 180.0;

        const vShader = new VertexShaderBuilder()
            .vertexInputData(true, false, true, true)
            .matrixData(true)
            .directionalLightEmbeded(
                [0.3, 0.3, 0.3], [1.0, 1.0, 1.0], [0.85, 0.8, 0.75])
            .exchangeData(false, true, true)
            .build();

        const fShader = new FragmentShaderBuilder()
            .exchangeData(false, true, true)
            .build();

        const prog = ShaderProgram.createProgram(gl, vShader, fShader);
        this.prog = prog;
        this.matrix = glMatrix.mat4.create();

        this.projMatrix = glMatrix.mat4.create();
        this.viewModelMatrix = glMatrix.mat4.create();
        this.normalMatrix = glMatrix.mat4.create();

        mat4.perspective(this.projMatrix, 45 * TO_RAD,
            gl.canvas.clientWidth / gl.canvas.clientHeight,
            0.1, 100.0
        );





        const coordinates = [
            // Front face
            -1.0, -1.0, 1.0, //
            1.0, -1.0, 1.0, //
            1.0, 1.0, 1.0, //
            -1.0, 1.0, 1.0, // Back face
            -1.0, -1.0, -1.0, //
            -1.0, 1.0, -1.0, //
            1.0, 1.0, -1.0, //
            1.0, -1.0, -1.0, // Top face
            -1.0, 1.0, -1.0, //
            -1.0, 1.0, 1.0, //
            1.0, 1.0, 1.0, //
            1.0, 1.0, -1.0, // Bottom face
            -1.0, -1.0, -1.0, //
            1.0, -1.0, -1.0, //
            1.0, -1.0, 1.0, //
            -1.0, -1.0, 1.0, // Right face
            1.0, -1.0, -1.0, //
            1.0, 1.0, -1.0, //
            1.0, 1.0, 1.0, //
            1.0, -1.0, 1.0, // Left face
            -1.0, -1.0, -1.0, //
            -1.0, -1.0, 1.0, //
            -1.0, 1.0, 1.0, //
            -1.0, 1.0, -1.0 //
        ];


        const indices = [
            0, 1, 2, /*        */ 0, 2, 3, // front
            4, 5, 6, /*        */ 4, 6, 7, // back
            8, 9, 10, /*       */ 8, 10, 11, // top
            12, 13, 14, /*     */ 12, 14, 15, // bottom
            16, 17, 18, /*     */ 16, 18, 19, // right
            20, 21, 22, /*     */ 20, 22, 23 // left
        ];

        const textCoords = [
            // Front
            0.0, 0.0, //
            1.0, 0.0, //
            1.0, 1.0, //
            0.0, 1.0, //
            // Back
            0.0, 0.0, //
            1.0, 0.0, //
            1.0, 1.0, //
            0.0, 1.0, //
            // Top
            0.0, 0.0, //
            1.0, 0.0, //
            1.0, 1.0, //
            0.0, 1.0, //
            // Bottom
            0.0, 0.0, //
            1.0, 0.0, //
            1.0, 1.0, //
            0.0, 1.0,
            // Right
            0.0, 0.0, //
            1.0, 0.0, //
            1.0, 1.0, //
            0.0, 1.0, //
            // Left
            0.0, 0.0, //
            1.0, 0.0, //
            1.0, 1.0, //
            0.0, 1.0 //

        ];

        const normals = [
            // Front
            0.0, 0.0, 1.0,
            0.0, 0.0, 1.0,
            0.0, 0.0, 1.0,
            0.0, 0.0, 1.0,

            // Back
            0.0, 0.0, -1.0,
            0.0, 0.0, -1.0,
            0.0, 0.0, -1.0,
            0.0, 0.0, -1.0,

            // Top
            0.0, 1.0, 0.0,
            0.0, 1.0, 0.0,
            0.0, 1.0, 0.0,
            0.0, 1.0, 0.0,

            // Bottom
            0.0, -1.0, 0.0,
            0.0, -1.0, 0.0,
            0.0, -1.0, 0.0,
            0.0, -1.0, 0.0,

            // Right
            1.0, 0.0, 0.0,
            1.0, 0.0, 0.0,
            1.0, 0.0, 0.0,
            1.0, 0.0, 0.0,

            // Left
            -1.0, 0.0, 0.0, -1.0, 0.0, 0.0, -1.0, 0.0, 0.0, -1.0, 0.0, 0.0
        ];

        gl.useProgram(prog.idProgram);
        this.idVao = GlUtils.createVAO(gl);
        GlUtils.prepareVertexData(gl, coordinates, null, textCoords, indices, normals);

        //this.idTexture = GlUtils.loadTextureDefault(gl, "/images/textures/235.jpg",
        this.idTexture = GlUtils.loadTextureDefault(gl, "img/textures/235.jpg",
            (gl, idTexture, image) => {
                gl.generateMipmap(gl.TEXTURE_2D);
            });
        //      this.idTexture = GlUtils.loadTextureDefault(gl, "/images/textures/235.jpg");

    }

    onDrawFrame(appOGL) {
        super.onDrawFrame(appOGL);
        const mat4 = glMatrix.mat4;
        const matrix = this.matrix;
        const viewModelMatrix = this.viewModelMatrix;
        const normalMatrix = this.normalMatrix;
        const gl = appOGL.gl;
        const prog = this.prog;
        const TO_RAD = Math.PI / 180.0;


        var rotY = this.rotY;
        var rotX = this.rotX;

        rotY += 1.5 * TO_RAD
        if (rotY > 2 * Math.PI) {
            rotY = rotY - 2 * Math.PI
        }
        rotX += 1 * TO_RAD
        if (rotX > 2 * Math.PI) {
            rotX = rotX - 2 * Math.PI
        }

        this.rotX = rotX;
        this.rotY = rotY;


        mat4.identity(viewModelMatrix);
        mat4.translate(viewModelMatrix, viewModelMatrix, [0.0, 0.0, -6.0]);
        mat4.rotate(viewModelMatrix, viewModelMatrix, rotX, [1, 0, 0]);
        mat4.rotate(viewModelMatrix, viewModelMatrix, rotY, [0, 1, 0]);
        mat4.multiply(matrix, this.projMatrix, viewModelMatrix);



        mat4.invert(normalMatrix, viewModelMatrix);
        mat4.transpose(normalMatrix, normalMatrix);

        gl.useProgram(prog.idProgram);


        gl.uniformMatrix4fv(
            gl.getUniformLocation(prog.idProgram, 'matrix'),
            false, matrix);

        gl.uniformMatrix4fv(
            gl.getUniformLocation(prog.idProgram, 'normalMatrix'),
            false, normalMatrix);




        gl.activeTexture(gl.TEXTURE0);
        gl.bindTexture(gl.TEXTURE_2D, this.idTexture);
        const sampleLocation = gl.getUniformLocation(prog.idProgram, 'uSampler');
        gl.uniform1i(sampleLocation, 0);


        gl.bindVertexArray(this.idVao);
        gl.enable(gl.CULL_FACE);
        gl.drawElements(gl.TRIANGLES, 36, gl.UNSIGNED_BYTE, 0);
    }

}

File: lwjgl/src/main/kotlin/com/darkraha/opengldemokt/renders/LightTexturedCubeRender.kt
package com.darkraha.opengldemokt.renders

import com.darkraha.opengldemokt.gl.AppOGL
import com.darkraha.opengldemokt.gl.GlUtils
import com.darkraha.opengldemokt.gl.ShaderProgram
import com.darkraha.opengldemokt.gl.shaders.FragmentShaderBuilder
import com.darkraha.opengldemokt.gl.shaders.VertexShaderBuilder
import org.joml.Matrix4f
import org.lwjgl.opengl.GL33.*

class LightTexturedCubeRender : Render() {
    private lateinit var matrix: Matrix4f
    private lateinit var projMatrix: Matrix4f
    private lateinit var viewModelMatrix: Matrix4f
    private lateinit var normalMatrix: Matrix4f
    private lateinit var prog: ShaderProgram
    private var rotY = 0f
    private var rotX = 0f
    private var idVao = 0
    private lateinit var ids: IntArray
    private var idTexture = 0

    override fun onSetup(appOGL: AppOGL) {
        setSurfaceSize(appOGL.width, appOGL.height)
        glClearColor(0.0f, 0.0f, 0.0f, 0.0f)
        glEnable(GL_DEPTH_TEST)
        val vertexShader = VertexShaderBuilder()
            .vertexInputData(true, false, true, true)
            .matrixData(true)
            .directionalLightEmbeded(
                floatArrayOf(0.3f, 0.3f, 0.3f),
                floatArrayOf(1f, 1f, 1f),
                floatArrayOf(0.85f, 0.8f, 0.75f)
            )
            .exchangeData(false, true, true)
            .build()
        val fragmentShader = FragmentShaderBuilder()
            .exchangeData(false, true, true)
            .build()
        prog = ShaderProgram(vertexShader, fragmentShader)




        glUseProgram(prog!!.idProgram)
        matrix = Matrix4f()
        projMatrix = Matrix4f()
        projMatrix!!.identity()
            .perspective(45 * TO_RAD, aspect, 1f, 100f)
        viewModelMatrix = Matrix4f()
        normalMatrix = Matrix4f()
        val coords = floatArrayOf( // Front face
            -1.0f, -1.0f, 1.0f,
            1.0f, -1.0f, 1.0f,
            1.0f, 1.0f, 1.0f,
            -1.0f, 1.0f, 1.0f,  // Back face
            -1.0f, -1.0f, -1.0f,
            -1.0f, 1.0f, -1.0f,
            1.0f, 1.0f, -1.0f,
            1.0f, -1.0f, -1.0f,  // Top face
            -1.0f, 1.0f, -1.0f,
            -1.0f, 1.0f, 1.0f,
            1.0f, 1.0f, 1.0f,
            1.0f, 1.0f, -1.0f,  // Bottom face
            -1.0f, -1.0f, -1.0f,
            1.0f, -1.0f, -1.0f,
            1.0f, -1.0f, 1.0f,
            -1.0f, -1.0f, 1.0f,  // Right face
            1.0f, -1.0f, -1.0f,
            1.0f, 1.0f, -1.0f,
            1.0f, 1.0f, 1.0f,
            1.0f, -1.0f, 1.0f,  // Left face
            -1.0f, -1.0f, -1.0f,
            -1.0f, -1.0f, 1.0f,
            -1.0f, 1.0f, 1.0f,
            -1.0f, 1.0f, -1.0f
        )
        val texCoords = floatArrayOf( // Front
            0.0f, 0.0f,  //
            1.0f, 0.0f,  //
            1.0f, 1.0f,  //
            0.0f, 1.0f,  //
            // Back
            0.0f, 0.0f,  //
            1.0f, 0.0f,  //
            1.0f, 1.0f,  //
            0.0f, 1.0f,  //
            // Top
            0.0f, 0.0f,  //
            1.0f, 0.0f,  //
            1.0f, 1.0f,  //
            0.0f, 1.0f,  //
            // Bottom
            0.0f, 0.0f,  //
            1.0f, 0.0f,  //
            1.0f, 1.0f,  //
            0.0f, 1.0f,  //
            // Right
            0.0f, 0.0f,  //
            1.0f, 0.0f,  //
            1.0f, 1.0f,  //
            0.0f, 1.0f,  //
            // Left
            0.0f, 0.0f,  //
            1.0f, 0.0f,  //
            1.0f, 1.0f,  //
            0.0f, 1.0f
        )
        val normals = floatArrayOf( // front
            0.0f, 0.0f, 1.0f,  //
            0.0f, 0.0f, 1.0f,  //
            0.0f, 0.0f, 1.0f,  //
            0.0f, 0.0f, 1.0f,  //
            // Back
            0.0f, 0.0f, -1.0f,  //
            0.0f, 0.0f, -1.0f,  //
            0.0f, 0.0f, -1.0f,  //
            0.0f, 0.0f, -1.0f,  //
            // Top
            0.0f, 1.0f, 0.0f,  //
            0.0f, 1.0f, 0.0f,  //
            0.0f, 1.0f, 0.0f,  //
            0.0f, 1.0f, 0.0f,  //
            // Bottom
            0.0f, -1.0f, 0.0f,  //
            0.0f, -1.0f, 0.0f,  //
            0.0f, -1.0f, 0.0f,  //
            0.0f, -1.0f, 0.0f,  //
            // Right
            1.0f, 0.0f, 0.0f,  //
            1.0f, 0.0f, 0.0f,  //
            1.0f, 0.0f, 0.0f,  //
            1.0f, 0.0f, 0.0f,  //
            // Left
            -1.0f, 0.0f, 0.0f,  //
            -1.0f, 0.0f, 0.0f,  //
            -1.0f, 0.0f, 0.0f,  //
            -1.0f, 0.0f, 0.0f //
        )
        val indices = byteArrayOf(
            0, 1, 2,  /*        */0, 2, 3,  // front
            4, 5, 6,  /*        */4, 6, 7,  // back
            8, 9, 10,  /*       */8, 10, 11,  // top
            12, 13, 14,  /*     */12, 14, 15,  // bottom
            16, 17, 18,  /*     */16, 18, 19,  // right
            20, 21, 22,  /*     */20, 22, 23
        )
        idTexture = GlUtils.loadTex2DResDefault(0, "/textures/235.jpg")
        glGenerateMipmap(idTexture)
        idVao = glGenVertexArrays()
        glBindVertexArray(idVao)
        ids = GlUtils.prepareVertexData(coords, null, texCoords, indices, normals)
    }

    override fun onDrawFrame(appOGL: AppOGL) {
        rotY += 1.5f * TO_RAD
        if (rotY > 2 * PI_F) {
            rotY = rotY - 2 * PI_F
        }
        rotX += 1f * TO_RAD
        if (rotX > 2 * PI_F) {
            rotX = rotX - 2 * PI_F
        }
        glClear(GL_COLOR_BUFFER_BIT or GL_DEPTH_BUFFER_BIT)
        val samplerLocation = glGetUniformLocation(prog!!.idProgram, "uSampler")
        glActiveTexture(GL_TEXTURE0)
        glBindTexture(GL_TEXTURE_2D, idTexture)
        glUniform1i(samplerLocation, 0)
        glUseProgram(prog!!.idProgram)
        viewModelMatrix!!.identity()
            .translate(0f, 0f, -6f)
            .rotateAffineXYZ(rotX, rotY, 0f)
        viewModelMatrix!!.invert(normalMatrix)
        normalMatrix!!.transpose()
        matrix!!.set(projMatrix).mul(viewModelMatrix)
        GlUtils.bindMatrix(prog!!.idProgram, matrix)
        GlUtils.bindNormalMatrix(prog!!.idProgram, normalMatrix)
        glBindVertexArray(idVao)
        glEnable(GL_CULL_FACE)
        glDrawElements(GL_TRIANGLES, 36, GL_UNSIGNED_BYTE, 0)
    }

    override fun onDispose(appOGL: AppOGL) {
        prog!!.dispose()
        glDeleteTextures(idTexture)
        glDeleteVertexArrays(idVao)
        glDeleteBuffers(ids)
    }
}
